### PR TITLE
Benchmark array contracts

### PR DESCRIPTION
**Pull request by fuzzypixelz**
Add benchmarks in the context of #764 so we can evaluate results.

### CODE REVIEW

**Review by yannham — CHANGES_REQUESTED** [review]
It seems to be missing the `arrays/sum.ncl` file at least, or am I missing something?

**Comment by fuzzypixelz** [comment]
Just a quick mental note.

We need benchmarks that specifically target

File: benches/arrays/sum.ncl [hunk 1]
```diff
@@ -0,0 +1,12 @@
+let sum
+  | Array Num -> Num
+  = let rec aux = fun step xs =>
```
Comment (yannham) [inline benches/arrays/sum.ncl:3]: If the goal of this bench is a case where a lot of array contracts are applied at each recursive call, I would not care about stack space/tail calls and write the naive sum function `sum [] = 0, sum x:xs = 1 + sum xs`. Here, the `Array Num` input contract is only applied on the call to `sum`, but not to actual recursive calls to `aux` (although there's still `array.length` which applies an array contract each times), I think.

File: benches/arrays/primes.ncl [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+let is_prime 
+  | doc "Returns true if the argument is a prime number."
+  = fun x => x > 1 && array.all (fun d => x % d != 0) (range 2 (x - 1))
+in
+
+let Prime = fun label value =>
+  if is_prime value then
+    value
+  else
+    contract.blame_with "%{string.from_num value} is not a prime number." label
```
Comment (yannham) [inline benches/arrays/primes.ncl:20]: Not very important for a benchmark, but the idiomatic way would be to write 

```nickel
let Prime = contract.from_predicate is_prime in
```

**Review by yannham — APPROVED** [review]

File: benches/arrays/primes.ncl (hunk 2 above)
Comment (fuzzypixelz) [reply benches/arrays/primes.ncl:20]: I was aware of that. But I thought the error message could not be customized. Would this make for a new library function? i.e. `contract.from_predicate_with`? (the name is not the strongest argument here :P)

**Comment by yannham** [comment]
@fuzzypixelz it's a bit harder to follow and to review when adding new commits to this PR, which has already been approved twice. What about dropping the last commit (I think it adds a new benchmark? Also the commit contains trailing git conflict markers), merge as it is once the CI is green, and open a new PR for other benchmarks?

**Comment by yannham** [comment]
Thanks! LGTM :)

**Pull request merged**